src/ifcgeom/taxonomy.h

```diff
@@ -125,6 +125,41 @@ struct item {
 	uint32_t identity() const { return identity_; }
 };
 
+struct implicit_item : public item {
+	DECLARE_PTR(implicit_item)
+	
+	virtual item::ptr evaluate() const = 0;
+};
+
+struct piecewise_function : public implicit_item {
+	DECLARE_PTR(piecewise_function)
+
+	std::vector<std::pair<double, std::function<Eigen::VectorXd(double u)>>> spans;
+
+	void print(std::ostream& o, int indent = 0) const {
+		o << "piecewise_function" << std::endl;
+	}
+
+	virtual piecewise_function* clone_() const { return new piecewise_function(*this); }
+	virtual kinds kind() const { return PIECEWISE_FUNCTION; }
+
+	virtual size_t calc_hash() const {
+		auto v = std::make_tuple(static_cast<size_t>(PIECEWISE_FUNCTION), 0);
+		return boost::hash<decltype(v)>{}(v);
+	}
+
+	virtual item::ptr evaluate() const;
+
+	Eigen::VectorXd evaluate(double u) const {
+		// @todo optimize, assume monotonic evaluation and store last evaluated segment?
+		for (auto& [length, fn] : spans) {
+			if (u < length) {
+				return fn(u);
+			}
+			u -= length;
+		}
```

Comment:
not all paths return a value.

Reply:
Hm yes indeed. I don't know what's reasonable here. I have also heard that sometimes vertical alignments actually extend beyond the horizontal in case of e.g linear horizontal segments and you'd be assumed to simply extrapolate the linear segment. Is that a thing? Should we accommodate for that? We could add a check whether it's the last segment and in case it's last don't check for `u  < length` that would also save us some difficulties in dealing with the last segment, because due to rounding errors, evaluating the last 0-length segment will likely never happen in our current case. The only case not handled is negative `u` then. I don't have an opinion what to do (either throw an exception or extrapolate first segment beyond it's start).

Reply:
Maybe also something for @civilx64 and the validation service. I can imagine at least a warning would be good to have in place in case of disagreement in range.

Reply:
I reviewed the spec and didn't find anything specifically addressing this.  I do remember seeing a use case or example specifically focusing on vertical range shorter than horizontal, which is common - and even recommended - in my professional experience.

I recommend the following for our implementation in IFCOS:

1. Vertical range can be shorter than horizontal at the beginning, end, or both limits of horizontal.
2. z values for horizontal (x, y) beyond the range of vertical shall be null
3. Vertical range beyond horizontal will be logged as a warning in IFCOS and the validation service.

Reply:
@civilx64 
> z values for horizontal (x, y) beyond the range of vertical shall be null

How is this accomplished? The result of evaluating IfcGradientCurve or IfcSegmentedReferenceCurve is a tuple xyz of real numbers.

Reply:
One option would be to use `std::numeric_limits<double>::quiet_NaN()`. It's maybe not very pretty, but should be supported everywhere since it's part of the IEEE spec.